# 1.0 Data Wrangling <a name="1.1_datawrangling"></a>


## Contents <a name="contents"></a>

* [1 Data Wrangling](#1.1_datawrangling)

    * [Contents](#contents)


* [1.2 Introduction](#1.2_introduction)

    * [1.2.1 Problem Recap](#1.2.1_recap)

    * [1.2.2 Data Wrangling Introduction](#1.2.2_wrangling)
    

* [1.3 Imports](#1.3_imports)


* [1.4 Load Data](#1.4_loaddata)


* [1.5 Fashion Dataset](#1.5_exploredata)

    * [1.5.1 Fashion data features](#1.5.1_feat)
    
    * [1.5.2 Fashion missing data](#1.5.2_f_missing)


* [1.6 Electronics dataset](#1.6_electronics)
    
    * [1.6.1 Electronics data features](#1.6.1_feat)

    * [1.6.2 Electronics missing data](#1.6.2_e_missing)
        

* [1.7 Home/kitchen dataset](#1.7_h/k)
    
    * [1.7.1 Home/kitchen data features](#1.7.1_h/k_exp)
        
    * [1.7.2 Home/kitchen missing data](#1.7.2_h/k_missing)
            
        
* [1.8 Save Data](#1.8)
    
    * [1.8.1 Creating a parsing function](#1.8.1)
    
    * [1.8.2 Fashion CSV](#1.8.2)
        
    * [1.8.3 Electronics CSV](#1.8.3)
        
    * [1.8.4 Home/Kitchen CSV](#1.8.4)


* [1.9 Summary](#1.9_summary)

## 1.1 Data Wrangling 

## 1.2 Introduction<a name="1.2_introduction"></a>
In this notebook, we are going to primarily be focusing on loading the data, examining missing and duplicated values, and dropping unneccesary features. We may do a small amount of data cleaning, but we do not want to prematurely drop features that we may later need.

### 1.2.1 Problem Recap <a name="1.2.1_recap"></a>

Using customer text data about amazon products, we will build, evaluate and compare models to estimate the probability that a given text review can be classified as “positive”, “neutral”, or “negative”.

Our goal is to build a text classifier using Amazon product review data which can be used to analyze customer sentiment which does not have accompanying numeric data.

### 1.2.2 Data Wrangling Introduction <a name="1.2.2_wrangling"></a>
We will read in the data, remove unnecessary columns/features, and then save it in an easier-to-work-with tabular format (CSV). 

## 1.3 Imports <a name="1.3_imports"></a>

In [1]:
import json
import csv
import pandas as pd

## 1.4 Load Data<a name="1.4_loaddata"></a>
Datasets are sourced from:

[Jianmo Ni, Jiacheng Li, Julian McAuley
Empirical Methods in Natural Language Processing (EMNLP), 2019](https://nijianmo.github.io/amazon/index.html)

The data is split up by product category. 

Our categories will be:

#### Fashion (883,636 reviews)

#### Electronics (20,994,353 reviews)

#### Home and Kitchen (21,928,568 reviews)

The data is available as compressed JSON files. We will use pandas to load the fashion data and determine which features to keep or drop and which features need preliminary cleaning.

Because of how large the Electronics and Home and Kitchen datasets are (approx. 10 GB each), we'll use the fashion dataset to explore the features and build a set of functions to parse each of the three category files. We will stream the JSON data, drop the values/features we do not want to keep, and save to separate CSV files.

In [2]:
!ls -hl data/raw

ls: cannot access 'data/raw': No such file or directory


## 1.5 Explore the Fashion data <a name="1.5_exploredata"></a>

Let's take a look at the top of the **Fashion** dataset

In [3]:
fashion_head = pd.read_json("../data/raw/AMAZON_FASHION.json", lines=True, nrows=200)

In [4]:
fashion_head.head(3)

overall  verified   reviewTime      reviewerID        asin reviewerName  \
0        5      True  10 20, 2014  A1D4G1SNUZWQOT  7106116521        Tracy   
1        2      True  09 28, 2014  A3DDWDH9PX2YX2  7106116521    Sonja Lau   
2        4     False  08 25, 2014  A2MWC41EW7XL15  7106116521     Kathleen   

                                          reviewText  \
0                             Exactly what I needed.   
1  I agree with the other review, the opening is ...   
2  Love these... I am going to order another pack...   

                                             summary  unixReviewTime  vote  \
0                             perfect replacements!!      1413763200   NaN   
1  I agree with the other review, the opening is ...      1411862400   3.0   
2                                My New 'Friends' !!      1408924800   NaN   

  style image  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN

In [5]:
fashion_head.dtypes

overall             int64
verified             bool
reviewTime         object
reviewerID         object
asin               object
reviewerName       object
reviewText         object
summary            object
unixReviewTime      int64
vote              float64
style              object
image              object
dtype: object

### 1.5.1 Fashion data features <a name="1.5.1_feat"></a>

The **overall** feature is the number of stars, which we'll be using for our categorical target feature (positive/neutral/negative). It's an integer from 1-5. 

The **verified** feature is a boolean and determines whether Amazon can confirm the user purchased the item they're reviewing (not a free/discounted reviewed item).

The **reviewTime** is in a date string format, and we'll drop it and keep the **unixReviewTime** instead.

The **reviewerID** and **reviewerName** have similar information and we can keep just one (the ID should be fine).

The **asin** is the specific product ID. We could keep it if we want to get item metadata later and use that for additional exploratory analysis (ie. price, product tags, product subcategory).

The **summary** is a user provided (or amazon generated) short item summary.

The **vote** column seems to indicate how many users found a review useful.

### Important features:

The **reviewText** will be the main input feature/features we will be using for our sentiment prediction and modelling.

Our target feature will be based on the **overall** column. We will be using the star value for each review to convert the data to Negative/Neutral/Positive categories. This will be the feature that our model trains and predicts with.

Let's check the **style** column.

In [6]:
fashion_head[~(fashion_head["style"].isna())]

overall  verified   reviewTime      reviewerID        asin  \
7          3      True  09 22, 2013  A1BB77SEBQT8VX  B00007GDFV   
8          3      True  07 17, 2013   AHWOW7D1ABO9C  B00007GDFV   
9          3      True  04 13, 2013   AKS3GULZE0HFC  B00007GDFV   
10         4      True   03 9, 2013   A38NS6NF6WPXS  B00007GDFV   
11         2      True  01 27, 2013  A1KOKO3HTSAI1H  B00007GDFV   
..       ...       ...          ...             ...         ...   
175        5      True  08 28, 2016  A3U8IMV1FKI5Y3  B00008JPRZ   
176        5      True  05 25, 2016  A21CTBULJC61N1  B00008JPRZ   
193        5     False   03 2, 2006  A3JGLFGCB7KZQX  B0000AWXMM   
194        4     False  08 22, 2005  A179HMPWBPTJ2I  B0000AWXMM   
195        5     False   08 1, 2005  A1LZHPFUUE6ALJ  B0000AWXMM   

           reviewerName                                         reviewText  \
7    Darrow H Ankrum II  mother - in - law wanted it as a present for h...   
8                rosieO  Item is of good quality. Looks great, too. But...   
9            M. Waltman  I had used my last el-cheapo fake leather ciga...   
10             BTDoxies  This brand has been around a long time and you...   
11         Robin Howard  I smoke 100's and these are NOT made for them....   
..                  ...                                                ...   
175                 SPB            Love shirt and even makes me look good!   
176       Harrycarnival                                        great shirt   
193                  DG  It was exactly as described.  Beautiful ring, ...   
194       Leana J. Kemp  This is a nice ring.  It is simple and elegant...   
195                  KK  it is a beautiful ring. very shiny and classy....   

                                               summary  unixReviewTime  vote  \
7                                  bought as a present      1379808000   NaN   
8                            Buxton heiress collection      1374019200   NaN   
9                       Top Clasp Broke Within 3 days!      1365811200   NaN   
10                                     BUXTON QUALITY!      1362787200   NaN   
11   Buxton Heiress Collection Black Leather Cigare...      1359244800   NaN   
..                                                 ...             ...   ...   
175                                  Love the shirt!!!      1472342400   NaN   
176                                         Five Stars      1464134400   NaN   
193           BEAUTIFUL RING!!!! NO REGRETS WHATSOEVER      1141257600   NaN   
194                                    White Gold Ring      1124668800   6.0   
195                                           loved it      1122854400   3.0   

                                               style image  
7                               {'Color:': ' Black'}   NaN  
8                               {'Color:': ' Black'}   NaN  
9                               {'Color:': ' Black'}   NaN  
10                              {'Color:': ' Black'}   NaN  
11                              {'Color:': ' Black'}   NaN  
..                                               ...   ...  
175      {'Size:': ' 16.5 - 34', 'Color:': ' White'}   NaN  
176      {'Size:': ' 17.5 - 37', 'Color:': ' White'}   NaN  
193  {'Size:': ' 12', 'Metal Type:': ' yellow-gold'}   NaN  
194   {'Size:': ' 10', 'Metal Type:': ' white-gold'}   NaN  
195  {'Size:': ' 9.5', 'Metal Type:': ' white-gold'}   NaN  

[167 rows x 12 columns]

The **style** column seems to have sizing and color info but is missing for most reviews.

### 1.5.2 Fashion missing data <a name="1.5.2_f_missing"></a>

In [7]:
fashion = pd.read_json("../data/raw/AMAZON_FASHION.json", lines=True, chunksize=100000)

Using a json_reader iterator from Pandas, we can iterate through the data rather than loading the entire dataset into memory


In [8]:
missing_stats=[]
for chunk in fashion:
    missing_stats.append(chunk.isna().sum())

In [9]:
missing_vals = pd.DataFrame(missing_stats)

In [10]:
missing_vals

overall  verified  reviewTime  reviewerID  asin  reviewerName  reviewText  \
0        0         0           0           0     0             6         109   
1        0         0           0           0     0            18         108   
2        0         0           0           0     0            13         173   
3        0         0           0           0     0            14         172   
4        0         0           0           0     0             6          93   
5        0         0           0           0     0             5         123   
6        0         0           0           0     0             7         174   
7        0         0           0           0     0            11         152   
8        0         0           0           0     0            12         129   

   summary  unixReviewTime   vote  style  image  
0       39               0  90595  28037  98423  
1       61               0  89490  49765  97805  
2       68               0  89152  44147  95963  
3       77               0  86690  41255  94941  
4       26               0  92429  78710  98412  
5       49               0  93244  88410  97702  
6       78               0  92668  87838  96291  
7       76               0  92517  87214  95966  
8       59               0  76951  73691  79326

In [11]:
missing_vals.sum()

overall                0
verified               0
reviewTime             0
reviewerID             0
asin                   0
reviewerName          92
reviewText          1233
summary              533
unixReviewTime         0
vote              803736
style             579067
image             854829
dtype: int64

We have no missing values in our **overall** column, which is excellent. The **vote**, **style** and **image** columns have a lot of missing values, which is fine as we'll be dropping those columns. There are not many missing **reviewText** values.

Our final column/feature list to keep will be:

* **overall**
* **reviewerID**
* **verified**
* **asin**
* **reviewText**
* **summary**
* **unixReviewTime**

In [12]:
cols_to_keep = ["overall",  "verified", "reviewerID", "asin", "reviewText", "summary", "unixReviewTime"]

Let's do a quick preview to make sure our column list is correct.

In [13]:
fashion_head[cols_to_keep].head()

overall  verified      reviewerID        asin  \
0        5      True  A1D4G1SNUZWQOT  7106116521   
1        2      True  A3DDWDH9PX2YX2  7106116521   
2        4     False  A2MWC41EW7XL15  7106116521   
3        2      True  A2UH2QQ275NV45  7106116521   
4        3     False   A89F3LQADZBS5  7106116521   

                                          reviewText  \
0                             Exactly what I needed.   
1  I agree with the other review, the opening is ...   
2  Love these... I am going to order another pack...   
3                                too tiny an opening   
4                                               Okay   

                                             summary  unixReviewTime  
0                             perfect replacements!!      1413763200  
1  I agree with the other review, the opening is ...      1411862400  
2                                My New 'Friends' !!      1408924800  
3                                          Two Stars      1408838400  
4                                        Three Stars      1406419200

## 1.6 Electronics dataset <a name="1.6_electronics"></a>


In [14]:
elec_head = pd.read_json("../data/raw/Electronics.json", lines=True, nrows=1000)

In [15]:
elec_head.head()

overall  verified   reviewTime      reviewerID      asin  \
0        5      True  07 17, 2002  A1N070NS9CJQ2I  60009810   
1        5     False   07 6, 2002  A3P0KRKOBQK1KN  60009810   
2        5     False   07 3, 2002  A192HO2ICJ75VU  60009810   
3        4     False  06 30, 2002  A2T278FKFL3BLT  60009810   
4        5     False  06 28, 2002  A2ZUXVTW8RXBXW  60009810   

                       style reviewerName  \
0  {'Format:': ' Hardcover'}   Teri Adams   
1  {'Format:': ' Hardcover'}     Willa C.   
2  {'Format:': ' Hardcover'}          Kit   
3  {'Format:': ' Hardcover'}       Andres   
4  {'Format:': ' Hardcover'}         John   

                                          reviewText  \
0  This was the first time I read Garcia-Aguilera...   
1  As with all of Ms. Garcia-Aguilera's books, I ...   
2  I've not read any of Ms Aguilera's works befor...   
3  This romance novel is right up there with the ...   
4  Carolina Garcia Aguilera has done it again.  S...   

                          summary  unixReviewTime  vote image  
0                   Hit The Spot!      1026864000   NaN   NaN  
1  one hot summer is HOT HOT HOT!      1025913600   NaN   NaN  
2                  One Hot Summer      1025654400   2.0   NaN  
3               I love this book!      1025395200   3.0   NaN  
4                    One Hot Book      1025222400   NaN   NaN

### 1.6.1 Electronics data features <a name="1.6.1_feat"></a>



In [16]:
elec_head.dtypes

overall             int64
verified             bool
reviewTime         object
reviewerID         object
asin                int64
style              object
reviewerName       object
reviewText         object
summary            object
unixReviewTime      int64
vote              float64
image              object
dtype: object

It looks like our column names and data is pretty similar. Let's check and make sure the columns we wish to keep have the same names.

In [17]:
elec_head[cols_to_keep]

overall  verified      reviewerID       asin  \
0          5      True  A1N070NS9CJQ2I   60009810   
1          5     False  A3P0KRKOBQK1KN   60009810   
2          5     False  A192HO2ICJ75VU   60009810   
3          4     False  A2T278FKFL3BLT   60009810   
4          5     False  A2ZUXVTW8RXBXW   60009810   
..       ...       ...             ...        ...   
995        5      True  A107S4MT25VXQ5  594481902   
996        4      True  A1T8TP4OHT5T13  594481902   
997        5      True  A3HICVLF4PFFMN  594481902   
998        5      True   AY0BL2JD0GAFT  594481902   
999        5      True  A2KK9GUTOXMB43  594481902   

                                            reviewText  \
0    This was the first time I read Garcia-Aguilera...   
1    As with all of Ms. Garcia-Aguilera's books, I ...   
2    I've not read any of Ms Aguilera's works befor...   
3    This romance novel is right up there with the ...   
4    Carolina Garcia Aguilera has done it again.  S...   
..                                                 ...   
995  Perfect...Nook brand so exactly what was neede...   
996  This cord helps me have two locations to plug ...   
997  bought for a spare for my 9" Nook HD and it fi...   
998  Lost my charger. Found this one for less than ...   
999  Needed a second Nook charger to leave in my tr...   

                            summary  unixReviewTime  
0                     Hit The Spot!      1026864000  
1    one hot summer is HOT HOT HOT!      1025913600  
2                    One Hot Summer      1025654400  
3                 I love this book!      1025395200  
4                      One Hot Book      1025222400  
..                              ...             ...  
995           Juat what I needed...      1399507200  
996                      Extra Cord      1399420800  
997                       great fit      1399248000  
998   Lost the.charger.for.my Nook!      1399248000  
999                  Second charger      1399075200  

[1000 rows x 7 columns]

### 1.6.2 Electronics missing data <a name=1.6.2_e_missing></a>

In [18]:
elect = pd.read_json("../data/raw/Electronics.json", lines=True, chunksize=100000)

Using a json_reader iterator from Pandas, we'll iterate through the dataset to get missing value summaries, like we did with the fashion dataset.

In [19]:
missing_stats=[]
for chunk in elect:
    missing_stats.append(chunk.isna().sum())

In [20]:
missing_vals = pd.DataFrame(missing_stats)

In [21]:
missing_vals.sum()

overall                  0
verified                 0
reviewTime               0
reviewerID               0
asin                     0
style             10497616
reviewerName          1713
reviewText            9684
summary               4754
unixReviewTime           0
vote              18300976
image             20645630
dtype: int64

Similarly to the fashion dataset, the style, vote and image columns have lots of missing values but we'll be dropping those columns. The two important columns, **overall** and **reviewText** have 0 and 9684 missing values respectively out of about 20 million reviews.

## 1.7 Home/kitchen dataset <a name="1.7_h/k"></a>            

In [22]:
hk_head = pd.read_json("../data/raw/Home_and_Kitchen.json", lines=True, nrows=200)

In [23]:
hk_head.head()

overall  vote  verified   reviewTime      reviewerID        asin  \
0        5   2.0      True  08 31, 2010  A3NSN9WOX8470M  0006564224   
1        5   2.0      True   04 2, 2010  A2AMX0AJ2BUDNV  0006564224   
2        5   NaN      True   11 5, 2015   A8LUWTIPU9CZB  0560467893   
3        4   4.0      True  10 29, 2015   AABKIIHAL0L66  0560467893   
4        3   NaN      True   09 9, 2015  A3DA0KIQ5OBK5C  0560467893   

    reviewerName                                         reviewText  \
0            mmm  I don't use these for their original use, and ...   
1  John R. Welch  Seems a bit expensive for a plastic bottle, bu...   
2   Linda Fahner                           Great product, love it!!   
3   TheBlueChain  This is a sturdy floating corner shelf!  We mo...   
4    angelaarden  I purchased 4 of these shelves. they look grea...   

                                        summary  unixReviewTime style image  
0                                  Many uses...      1283212800   NaN   NaN  
1                              Dispenser bottle      1270166400   NaN   NaN  
2                                    Five Stars      1446681600   NaN   NaN  
3  Sturdy Shelf, Poor Installation Instructions      1446076800   NaN   NaN  
4                   Look great - one bad one...      1441756800   NaN   NaN

In [24]:
hk_head.dtypes

overall             int64
vote              float64
verified             bool
reviewTime         object
reviewerID         object
asin               object
reviewerName       object
reviewText         object
summary            object
unixReviewTime      int64
style              object
image              object
dtype: object

### 1.7.1 Home/kitchen data features <a name="1.7.1_h/k_exp"></a>

In [25]:
hk_head

overall  vote  verified   reviewTime      reviewerID        asin  \
0          5   2.0      True  08 31, 2010  A3NSN9WOX8470M  0006564224   
1          5   2.0      True   04 2, 2010  A2AMX0AJ2BUDNV  0006564224   
2          5   NaN      True   11 5, 2015   A8LUWTIPU9CZB  0560467893   
3          4   4.0      True  10 29, 2015   AABKIIHAL0L66  0560467893   
4          3   NaN      True   09 9, 2015  A3DA0KIQ5OBK5C  0560467893   
..       ...   ...       ...          ...             ...         ...   
195        5   NaN      True   09 1, 2015   A1PNQ4F0UC9V4  1605161810   
196        5   NaN      True  08 14, 2015   AC1W8G7HU9VH8  1605161810   
197        5   NaN      True  07 18, 2015   AFY118HH4OG7F  1605161810   
198        5   NaN      True  06 19, 2015   AG6BH7V520TJ5  1605161810   
199        5   NaN      True  06 13, 2015  A3MNFE8HZDJ3WC  1605161810   

         reviewerName                                         reviewText  \
0                 mmm  I don't use these for their original use, and ...   
1       John R. Welch  Seems a bit expensive for a plastic bottle, bu...   
2        Linda Fahner                           Great product, love it!!   
3        TheBlueChain  This is a sturdy floating corner shelf!  We mo...   
4         angelaarden  I purchased 4 of these shelves. they look grea...   
..                ...                                                ...   
195             party                         I like this and thank you.   
196    Jeanette Dodds  My mom loves it! Delivery was exactly on time....   
197    GothicPoohBear                                      good quality!   
198               Deb                                         Beautiful!   
199  Madeline Collins  Perfect magnet for my fridge.  The picture ins...   

                                          summary  unixReviewTime style  \
0                                    Many uses...      1283212800   NaN   
1                                Dispenser bottle      1270166400   NaN   
2                                      Five Stars      1446681600   NaN   
3    Sturdy Shelf, Poor Installation Instructions      1446076800   NaN   
4                     Look great - one bad one...      1441756800   NaN   
..                                            ...             ...   ...   
195                                    Five Stars      1441065600   NaN   
196                        Five Stars! Beautiful!      1439510400   NaN   
197                                  good quality      1437177600   NaN   
198                                      Love it!      1434672000   NaN   
199                                  Great magnet      1434153600   NaN   

                                                 image  
0                                                  NaN  
1                                                  NaN  
2                                                  NaN  
3                                                  NaN  
4                                                  NaN  
..                                                 ...  
195                                                NaN  
196  [https://images-na.ssl-images-amazon.com/image...  
197                                                NaN  
198                                                NaN  
199                                                NaN  

[200 rows x 12 columns]

Looks like the same columns names. Let's make sure.

In [26]:
hk_head[cols_to_keep]

overall  verified      reviewerID        asin  \
0          5      True  A3NSN9WOX8470M  0006564224   
1          5      True  A2AMX0AJ2BUDNV  0006564224   
2          5      True   A8LUWTIPU9CZB  0560467893   
3          4      True   AABKIIHAL0L66  0560467893   
4          3      True  A3DA0KIQ5OBK5C  0560467893   
..       ...       ...             ...         ...   
195        5      True   A1PNQ4F0UC9V4  1605161810   
196        5      True   AC1W8G7HU9VH8  1605161810   
197        5      True   AFY118HH4OG7F  1605161810   
198        5      True   AG6BH7V520TJ5  1605161810   
199        5      True  A3MNFE8HZDJ3WC  1605161810   

                                            reviewText  \
0    I don't use these for their original use, and ...   
1    Seems a bit expensive for a plastic bottle, bu...   
2                             Great product, love it!!   
3    This is a sturdy floating corner shelf!  We mo...   
4    I purchased 4 of these shelves. they look grea...   
..                                                 ...   
195                         I like this and thank you.   
196  My mom loves it! Delivery was exactly on time....   
197                                      good quality!   
198                                         Beautiful!   
199  Perfect magnet for my fridge.  The picture ins...   

                                          summary  unixReviewTime  
0                                    Many uses...      1283212800  
1                                Dispenser bottle      1270166400  
2                                      Five Stars      1446681600  
3    Sturdy Shelf, Poor Installation Instructions      1446076800  
4                     Look great - one bad one...      1441756800  
..                                            ...             ...  
195                                    Five Stars      1441065600  
196                        Five Stars! Beautiful!      1439510400  
197                                  good quality      1437177600  
198                                      Love it!      1434672000  
199                                  Great magnet      1434153600  

[200 rows x 7 columns]

### 1.7.2 Home/kitchen missing <a name="1.7.2_h/k_missing"></a>

In [27]:
hk = pd.read_json("../data/raw/Home_and_Kitchen.json", lines=True, chunksize=100000)

Using a json_reader iterator from Pandas, we'll iterate through the dataset to get missing value summaries, like we did with the fashion dataset.

In [28]:
missing_stats=[]
for chunk in hk:
    missing_stats.append(chunk.isna().sum())

In [29]:
missing_vals = pd.DataFrame(missing_stats)

In [30]:
missing_vals.sum()

overall                  0
vote              19117548
verified                 0
reviewTime               0
reviewerID               0
asin                     0
reviewerName          1160
reviewText           14866
summary               5781
unixReviewTime           0
style              9785019
image             21296935
dtype: int64

Out of about 22 million reviews, we have no missing **overall** ratings and only 14,866 missing **reviewText**.

## 1.8 Save Data <a name="1.8"></a>

### 1.8.1 Creating a parsing function <a name="1.8.1"></a>

In [31]:
def df_cleaner(temp_df_chunk: pd.DataFrame,
               columns_to_keep=cols_to_keep) -> pd.DataFrame:
    
    """This function will take in a chunk of our dataset as a dataframe.
       It will return only the columns we wish.
       
       Takes:
           temp_df_chunk: a dataframe to be edited
           columns_to_keep: the list of columns we wish to keep from that df          
       Returns:
           clean_df: a smaller df with drop_by_row NaNs dropped and only a subset of original columns
       """
  
    return temp_df_chunk[columns_to_keep]

Let's do a quick test.

In [32]:
df_cleaner(hk_head)

overall  verified      reviewerID        asin  \
0          5      True  A3NSN9WOX8470M  0006564224   
1          5      True  A2AMX0AJ2BUDNV  0006564224   
2          5      True   A8LUWTIPU9CZB  0560467893   
3          4      True   AABKIIHAL0L66  0560467893   
4          3      True  A3DA0KIQ5OBK5C  0560467893   
..       ...       ...             ...         ...   
195        5      True   A1PNQ4F0UC9V4  1605161810   
196        5      True   AC1W8G7HU9VH8  1605161810   
197        5      True   AFY118HH4OG7F  1605161810   
198        5      True   AG6BH7V520TJ5  1605161810   
199        5      True  A3MNFE8HZDJ3WC  1605161810   

                                            reviewText  \
0    I don't use these for their original use, and ...   
1    Seems a bit expensive for a plastic bottle, bu...   
2                             Great product, love it!!   
3    This is a sturdy floating corner shelf!  We mo...   
4    I purchased 4 of these shelves. they look grea...   
..                                                 ...   
195                         I like this and thank you.   
196  My mom loves it! Delivery was exactly on time....   
197                                      good quality!   
198                                         Beautiful!   
199  Perfect magnet for my fridge.  The picture ins...   

                                          summary  unixReviewTime  
0                                    Many uses...      1283212800  
1                                Dispenser bottle      1270166400  
2                                      Five Stars      1446681600  
3    Sturdy Shelf, Poor Installation Instructions      1446076800  
4                     Look great - one bad one...      1441756800  
..                                            ...             ...  
195                                    Five Stars      1441065600  
196                        Five Stars! Beautiful!      1439510400  
197                                  good quality      1437177600  
198                                      Love it!      1434672000  
199                                  Great magnet      1434153600  

[200 rows x 7 columns]

### 1.8.2 Fashion CSV <a name="1.8.2"></a>

In [33]:
fashion = pd.read_json("../data/raw/AMAZON_FASHION.json", lines=True, chunksize=1_000_000)

In [34]:
for i, chunk in enumerate(fashion):
    df_cleaner(chunk).to_parquet(path=f"../data/edited/fashion{i}.parquet")

In [35]:
ls ../data/edited/

fashion0.parquet


In [36]:
pd.read_parquet("../data/edited/fashion0.parquet").head()

overall  verified      reviewerID        asin  \
0        5      True  A1D4G1SNUZWQOT  7106116521   
1        2      True  A3DDWDH9PX2YX2  7106116521   
2        4     False  A2MWC41EW7XL15  7106116521   
3        2      True  A2UH2QQ275NV45  7106116521   
4        3     False   A89F3LQADZBS5  7106116521   

                                          reviewText  \
0                             Exactly what I needed.   
1  I agree with the other review, the opening is ...   
2  Love these... I am going to order another pack...   
3                                too tiny an opening   
4                                               Okay   

                                             summary  unixReviewTime  
0                             perfect replacements!!      1413763200  
1  I agree with the other review, the opening is ...      1411862400  
2                                My New 'Friends' !!      1408924800  
3                                          Two Stars      1408838400  
4                                        Three Stars      1406419200

#### Looks like our function and loop worked great. It will take a lot longer for the two larger files.

### 1.8.3 Electronics CSV <a name="1.8.3"></a> 
    

In [37]:
%reset -f

In [38]:
import json
import csv
import pandas as pd

In [39]:
cols_to_keep = ["overall",  "verified", "reviewerID", "asin", "reviewText", "summary", "unixReviewTime"]

def df_cleaner(temp_df_chunk: pd.DataFrame,
               columns_to_keep=cols_to_keep) -> pd.DataFrame:
    return temp_df_chunk[columns_to_keep]

In [40]:
elect = pd.read_json("../data/raw/Electronics.json", lines=True, chunksize=1_000_000)

In [41]:
for i, chunk in enumerate(elect):
    df_cleaner(chunk).to_parquet(path=f"../data/edited/electronics{i}.parquet")

In [42]:
ls ../data/edited/

electronics0.parquet   electronics17.parquet  electronics5.parquet
electronics10.parquet  electronics18.parquet  electronics6.parquet
electronics11.parquet  electronics19.parquet  electronics7.parquet
electronics12.parquet  electronics1.parquet   electronics8.parquet
electronics13.parquet  electronics20.parquet  electronics9.parquet
electronics14.parquet  electronics2.parquet   fashion0.parquet
electronics15.parquet  electronics3.parquet
electronics16.parquet  electronics4.parquet


In [43]:
pd.read_parquet("../data/edited/electronics10.parquet").head(20)

overall  verified      reviewerID        asin  \
10000000        5      True  A2744822IV8IQ0  B00FOKZVPU   
10000001        1      True  A1T74DO8W3GESF  B00FOKZVPU   
10000002        2      True  A31R5W0381P1D3  B00FOKZVPU   
10000003        5      True  A2E6PCARIX47TW  B00FOKZVPU   
10000004        5      True  A1YB0WHSNT8BDY  B00FOKZVPU   
10000005        5      True  A224GZRA2FOIIM  B00FOKZVPU   
10000006        5      True  A3ARDA4ZMK7IR8  B00FOKZVNC   
10000007        4      True  A1IJHUGNRGM90H  B00FOKZVNC   
10000008        5      True  A394LMFTSQWP63  B00FOKZVNC   
10000009        1     False  A1JP5WW6I4JECX  B00FOKZVNC   
10000010        5      True   AQFY8PCPJY6A4  B00FOKZVNC   
10000011        5      True  A3NQ2O8CDLI0IY  B00FOKZVNC   
10000012        4      True   ASV5W8Q18GY2T  B00FOKZVNC   
10000013        2      True  A3RF9CUMFCD15Q  B00FOKZVNC   
10000014        5      True  A1VIKSS1TUP7BM  B00FOKZVNC   
10000015        4     False  A31G1OYVWM2590  B00FOKZVNC   
10000016        5      True  A2Y1E03V4CPTB7  B00FOKZVNC   
10000017        5      True  A2080NZFECK5FB  B00FOKZVNC   
10000018        5      True  A26FEZKH53AH7Z  B00FOKZVNC   
10000019        5      True   AQLOX1Z4NBYYY  B00FOKZVNC   

                                                 reviewText  \
10000000            My wife just loves it and was thrilled.   
10000001           Not the one I needed so it was returned.   
10000002  The openings are. Of placed to accommodate the...   
10000003                                        It's fits !   
10000004                                       As described   
10000005  The item is just as described and works wonder...   
10000006  My other Fintie was starting to look shabby af...   
10000007                                    Classy looking.   
10000008                                           Love it.   
10000009  This is a really nice looking but its not func...   
10000010  its soft and supple but is still sturdy enough...   
10000011                    Love the look, the ease of use.   
10000012  A little bit of a tight fit for the Kindle Fir...   
10000013  I cannot give a good review on something that ...   
10000014  Excellent item. Fits the Kindle it says it wil...   
10000015  Good case....safely protects my Fire.  The pen...   
10000016  I didn't know whether or not to expect a good-...   
10000017                    exactly what I was looking for.   
10000018              Beautiful  case more than  I expected   
10000019                    Again didn't fit my kindle fire   

                                                    summary  unixReviewTime  
10000000                                         Five Stars      1447027200  
10000001                         Fire HD 7 Case (old Model)      1445817600  
10000002  The openings are. Of placed to accommodate the...      1445558400  
10000003                                        Looks great      1443139200  
10000004                                         Five Stars      1442275200  
10000005                                      Great product      1441584000  
10000006  ... other Fintie was starting to look shabby a...      1498003200  
10000007                                         Four Stars      1497484800  
10000008                                         Five Stars      1496707200  
10000009  This is a really nice looking but its not func...      1493424000  
10000010                                          I love it      1491350400  
10000011                                         Five Stars      1491264000  
10000012                                But it's very nice.      1490659200  
10000013           Does not fit Kindle Fir 5th Gen right...      1489622400  
10000014                               Finally a good case!      1489276800  
10000015                                Kindle well covered      1489190400  
10000016                             Great quality and feel      1488931200  
10000017                                  

### 1.8.4 Home/Kitchen CSV <a name="1.8.4"></a>

In [44]:
hk = pd.read_json("../data/raw/Home_and_Kitchen.json", lines=True, chunksize=1_000_000)

In [46]:
for i, chunk in enumerate(hk):
    df_cleaner(chunk).to_parquet(path=f"../data/edited/kitchen{i}.parquet")

In [48]:
pd.read_parquet("../data/edited/kitchen10.parquet").head(20)

overall  verified      reviewerID        asin  \
11000000        5      True  A2JZREXUGDD9ME  B00K8NKYSC   
11000001        5      True  A1ECWTOCBGBC3L  B00K8NKYSC   
11000002        5      True  A3DRCC09NHCCU1  B00K8NKYSC   
11000003        5      True  A3TI1M63DXABS9  B00K8NKYSC   
11000004        5      True  A34T5IZTXB956R  B00K8NKYSC   
11000005        5      True   A6FEU436W333L  B00K8NKYSC   
11000006        5      True   AO7VAMMX2MBEG  B00K8NKYSC   
11000007        5      True  A12TX9Q29R3J26  B00K8NKYSC   
11000008        5      True  A19TEAU5UU0FCE  B00K8NKYSC   
11000009        5      True  A30JVFW7UK9HWU  B00K8NKYSC   
11000010        4      True   AYP3CJG8IGWNX  B00K8NKYPU   
11000011        4      True  A3UZJVSKXM1UEC  B00K8NKYPU   
11000012        5      True   AN9N1Q9KU3WWS  B00K8NKYPU   
11000013        5     False  A2RZRXIOMV1OA7  B00K8NKYPU   
11000014        5      True   A6V9CAQH5ZZPQ  B00K8NKYPU   
11000015        4      True  A1ZKG9J98KZBF6  B00K8NKYPU   
11000016        1      True  A1PA35VCM0SPT7  B00K8NKYPU   
11000017        5      True  A3CEOYC1J3GMRC  B00K8NKYPU   
11000018        5      True  A2V8F3ITUZTS8Q  B00K8NKYPU   
11000019        5      True  A1VEAR5ALMAGXL  B00K8NKYPU   

                                                 reviewText  \
11000000  Amazing, I always use this for my smoothies in...   
11000001  I was very excited to purchase the certified r...   
11000002  Wow! What a great machine. Wish I'd bought one...   
11000003  I'm so glad I got the model with the pre-progr...   
11000004  Smoothies, soups, spreads, raw deserts, ice cr...   
11000005         Had it over a year and it still work great   
11000006                                 Loving my Vitamix!   
11000007  i cannot say enough about how much I love my V...   
11000008  Waaayyy so much better than I even thought. Ma...   
11000009  Great machine!  We're having fun trying out th...   
11000010                Goods are good, the use of general.   
11000011  I am nt happen with this  item. After one minu...   
11000012  After some review and research, we decided to ...   
11000013  Have had my Vitamix for about 8 years.  Just r...   
11000014  The reviews don't do this justice.  Let me tel...   
11000015  It smells like the motor is burning up sometim...   
11000016  Everyone I know loves their Vitamix; so the Vi...   
11000017                                         Loving it!   
11000018  Shipped quickly enough.  Nicely packed.  New p...   
11000019  First off, I did get a pitcher that was making...   

                                                    summary  unixReviewTime  
11000000                                        A Necessity      1450396800  
11000001                                     Great blender!      1450051200  
11000002                                           Love It!      1449014400  
11000003                                    Awesome Blender      1448928000  
11000004  we make with this great product. Changed eatin...      1448323200  
11000005                                         Five Stars      1447977600  
11000006                                         Five Stars      1447718400  
11000007  i cannot say enough about how much I love my V...      1447286400  
11000008                                         My Vitamix      1446681600  
11000009             Start drinking your spinach right now!      1446595200  
11000010                                               GOOD      1438473600  
11000011       I am nt happen with this item. After one ...      1438473600  
11000012                    Save a couple hundred bucks....      1438473600  
11000013  It is still the best blender I've ever owned a...      1438473600  
11000014                        Reviews Don't do it Justice      1438387200  
11000015                                         Four Stars      1438300800  
11000016    Black Specs, Burning Smell- Totally Unimpressed      1438214400  
11000017                                  

In [49]:
!ls -ahl ../data/edited

total 7.2G
drwxrwxr-x 3 paul paul 4.0K Apr 14 11:23 .
drwxrwxr-x 5 paul paul 4.0K Apr  5 10:29 ..
-rw-rw-r-- 1 paul paul 245M Apr 14 11:18 electronics0.parquet
-rw-rw-r-- 1 paul paul 178M Apr 14 11:19 electronics10.parquet
-rw-rw-r-- 1 paul paul 163M Apr 14 11:19 electronics11.parquet
-rw-rw-r-- 1 paul paul 166M Apr 14 11:19 electronics12.parquet
-rw-rw-r-- 1 paul paul 177M Apr 14 11:19 electronics13.parquet
-rw-rw-r-- 1 paul paul 171M Apr 14 11:19 electronics14.parquet
-rw-rw-r-- 1 paul paul 172M Apr 14 11:20 electronics15.parquet
-rw-rw-r-- 1 paul paul 173M Apr 14 11:20 electronics16.parquet
-rw-rw-r-- 1 paul paul 214M Apr 14 11:20 electronics17.parquet
-rw-rw-r-- 1 paul paul 204M Apr 14 11:20 electronics18.parquet
-rw-rw-r-- 1 paul paul 158M Apr 14 11:20 electronics19.parquet
-rw-rw-r-- 1 paul paul 260M Apr 14 11:18 electronics1.parquet
-rw-rw-r-- 1 paul paul 159M Apr 14 11:20 electronics20.parquet
-rw-rw-r-- 1 paul paul 256M Apr 14 11:18 electronics2.parquet
-rw-rw-r-- 1 paul paul 

## 1.9 Summary <a name="1.9_summary"></a>

We have cleaned the three datasets. We did not drop any rows from our data yet, as we only had a small number of missing **reviewText** data, which we will explore in the future.

We dropped the **reviewTime** and **reviewerName** columns as they're somewhat duplicated by our **unixReviewTime** and **reviewerID** variables. We dropped the **vote**, **style**, and **image** columns, as they're both largely filled with missing values and unimportant to our analysis generally.

Feature variable: **reviewText** will be our main data source for analysis, though the review summary also contains text, possibly from the user, about their feelings about the item.

Target variable: **overall** stars (1-5) will be converted in the next part of the project into categorical variables, to predict positive, neutral and negative feelings regarding user's opinions of items.